File: typescript/src/realtime.ts
/**
 * Modei Realtime Adapter
 *
 * Handles Supabase Realtime connection for agent presence and verification.
 * This module is opt-in: only activated when realtime.enabled is true in config.
 *
 * Features:
 * - Exchange passport for short-lived Realtime JWT
 * - Track Presence with agent/passport metadata
 * - Respond to verify.request broadcasts with signed nonce
 * - Token refresh every 4 minutes
 * - Exponential backoff reconnection
 * - Cleanup on process exit
 */

import { createClient as createSupabaseClient, type SupabaseClient, type RealtimeChannel } from '@supabase/supabase-js';
import * as crypto from 'crypto';

export interface RealtimeConfig {
  enabled: boolean;
  supabaseUrl: string;
  tokenEndpoint: string;
}

export interface AgentCredentials {
  agentId: string;
  passportId: string;
  passportFingerprint: string;
  /** Base64-encoded Ed25519 private key for signing */
  privateKey: string;
  /** Base64-encoded Ed25519 public key */
  publicKey: string;
}

export interface RealtimeAdapter {
  connect(): Promise<void>;
  disconnect(): void;
}

const VERSION = '1.0.0';

/**
 * Create a Realtime adapter for agent presence and verification.
 */
export function createRealtimeAdapter(
  config: RealtimeConfig,
  credentials: AgentCredentials,
  apiKey: string,
): RealtimeAdapter {
  let supabase: SupabaseClient | null = null;
  let channel: RealtimeChannel | null = null;
  let refreshInterval: ReturnType<typeof setInterval> | null = null;
  let startedAt: string = new Date().toISOString();

  async function exchangeForToken(): Promise<string> {
    // Generate a challenge and sign it
    const challenge = crypto.randomBytes(32).toString('hex');
    const timestamp = new Date().toISOString();

    // Sign the challenge with the passport private key
    // Using dynamic import to avoid hard dependency on @noble/ed25519
    const ed = await import('@noble/ed25519');

    // Set up sha512 if needed
    if (!ed.hashes.sha512) {
      const { sha512 } = await import('@noble/hashes/sha2.js');
      ed.hashes.sha512 = sha512;
    }

    const privateKeyBytes = Buffer.from(credentials.privateKey, 'base64');
    const challengeBytes = Buffer.from(challenge, 'utf8');
    const signature = await ed.signAsync(challengeBytes, privateKeyBytes);
    const signatureHex = Buffer.from(signature).toString('hex');

    const response = await fetch(config.tokenEndpoint, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        agent_id: credentials.agentId,
        passport_id: credentials.passportId,
        challenge,
        signature: signatureHex,
        timestamp,
      }),
    });

    if (!response.ok) {
      const err = await response.json().catch(() => ({ error: 'Unknown error' }));
      throw new Error(`Failed to exchange passport for token: ${err.error || response.status}`);
    }

    const data = await response.json();
    return data.token;
  }

  async function signNonce(nonce: string): Promise<string> {
    const ed = await import('@noble/ed25519');
    if (!ed.hashes.sha512) {
      const { sha512 } = await import('@noble/hashes/sha2.js');
      ed.hashes.sha512 = sha512;
    }
    const privateKeyBytes = Buffer.from(credentials.privateKey, 'base64');
    const nonceBytes = Buffer.from(nonce, 'utf8');
    const signature = await ed.signAsync(nonceBytes, privateKeyBytes);
    return Buffer.from(signature).toString('hex');
  }

  async function connect() {
    startedAt = new Date().toISOString();

    // Step 1: Exchange passport for Realtime token
    const token = await exchangeForToken();
    console.error(`[modei-realtime] Token acquired for agent ${credentials.agentId}`);

    // Step 2: Connect to Supabase Realtime
    supabase = createSupabaseClient(config.supabaseUrl, token, {
      auth: { autoRefreshToken: false, persistSession: false },
    });

    supabase.realtime.setAuth(token);

    const channelName = `agent:${credentials.agentId}`;
    channel = supabase.channel(channelName, {
      config: { presence: { key: credentials.agentId } },
    });

    // Step 3: Listen for verify.request broadcasts
    channel.on('broadcast', { event: 'verify.request' }, async (payload) => {
      try {
        const { nonce, timestamp } = payload.payload;
        const signature = await signNonce(nonce);

        channel!.send({
          type: 'broadcast',
          event: 'verify.response',
          payload: {
            nonce,
            timestamp: new Date().toISOString(),
            uptime_seconds: Math.floor(process.uptime()),
            passport_fingerprint: credentials.passportFingerprint,
            sdk_version: `modei-mcp@${VERSION}`,
            signature,
          },
        });
        console.error(`[modei-realtime] Responded to verify challenge`);
      } catch (err) {
        console.error(`[modei-realtime] Failed to respond to verify challenge:`, err);
      }
    });

    // Step 4: Subscribe and track Presence
    channel.subscribe(async (status) => {
      if (status === 'SUBSCRIBED') {
        await channel!.track({
          agent_id: credentials.agentId,
          passport_id: credentials.passportId,
          passport_fingerprint: credentials.passportFingerprint,
          sdk: 'modei-mcp',
          sdk_version: VERSION,
          runtime: 'nodejs',
          env: process.env.NODE_ENV || 'development',
          started_at: startedAt,
        });
        console.error(`[modei-realtime] Presence tracked on channel ${channelName}`);
      }
    });

    // Step 5: Token refresh every 4 minutes
    refreshInterval = setInterval(async () => {
      try {
        const newToken = await exchangeForToken();
        supabase!.realtime.setAuth(newToken);
        console.error(`[modei-realtime] Token refreshed`);
      } catch (err) {
        console.error(`[modei-realtime] Token refresh failed:`, err);
      }
    }, 4 * 60 * 1000);

    // Step 6: Cleanup on process exit
    const cleanup = () => {
      disconnect();
      process.exit(0);
    };
    process.on('SIGTERM', cleanup);
    process.on('SIGINT', cleanup);
  }

  function disconnect() {
    if (refreshInterval) {
      clearInterval(refreshInterval);
      refreshInterval = null;
    }
    if (channel) {
      channel.untrack();
      channel.unsubscribe();
      channel = null;
    }
    if (supabase) {
      supabase.removeAllChannels();
      supabase = null;
    }
    console.error(`[modei-realtime] Disconnected`);
  }

  return { connect, disconnect };
}

/**
 * Compute a SHA-256 fingerprint of the passport public key.
 */
export function computeFingerprint(publicKeyBase64: string): string {
  const hash = crypto.createHash('sha256').update(Buffer.from(publicKeyBase64, 'base64')).digest('hex');
  return `sha256:${hash}`;
}

File: python/src/modei/realtime.py
"""
Modei Realtime Adapter (Python)

Handles Supabase Realtime connection for agent presence and verification.
This module is opt-in: install with `pip install modei-sdk[realtime]`.

Features:
- Exchange passport for short-lived Realtime JWT
- Track Presence with agent/passport metadata
- Respond to verify.request broadcasts with signed nonce
- Token refresh every 4 minutes
- Cleanup on exit
"""

from __future__ import annotations

import asyncio
import hashlib
import json
import os
import signal
import time
from dataclasses import dataclass
from typing import Any, Optional

import httpx

try:
    from nacl.signing import SigningKey
    from supabase import create_client, Client as SupabaseClient
except ImportError:
    raise ImportError(
        "Realtime support requires additional dependencies. "
        "Install with: pip install modei-sdk[realtime]"
    )

VERSION = "1.0.0"


@dataclass
class RealtimeConfig:
    """Configuration for Realtime connection."""
    enabled: bool = False
    supabase_url: str = ""
    token_endpoint: str = ""


@dataclass
class AgentCredentials:
    """Agent passport credentials for Realtime auth."""
    agent_id: str = ""
    passport_id: str = ""
    passport_fingerprint: str = ""
    private_key_base64: str = ""  # Ed25519 private key (base64)
    public_key_base64: str = ""   # Ed25519 public key (base64)


def compute_fingerprint(public_key_base64: str) -> str:
    """Compute SHA-256 fingerprint of the passport public key."""
    import base64
    pub_bytes = base64.b64decode(public_key_base64)
    digest = hashlib.sha256(pub_bytes).hexdigest()
    return f"sha256:{digest}"


class RealtimeAdapter:
    """
    Manages Supabase Realtime connection for agent presence and verification.

    Usage::

        config = RealtimeConfig(
            enabled=True,
            supabase_url="https://xyz.supabase.co",
            token_endpoint="https://modeitrust.ai/api/auth/realtime-token",
        )
        credentials = AgentCredentials(
            agent_id="agent_abc123",
            passport_id="pass_xyz",
            passport_fingerprint=compute_fingerprint(public_key_b64),
            private_key_base64="...",
            public_key_base64="...",
        )
        adapter = RealtimeAdapter(config, credentials, api_key="mod_live_...")
        await adapter.connect()
    """

    def __init__(
        self,
        config: RealtimeConfig,
        credentials: AgentCredentials,
        api_key: str,
    ):
        self.config = config
        self.credentials = credentials
        self.api_key = api_key
        self._supabase: Optional[SupabaseClient] = None
        self._channel: Any = None
        self._refresh_task: Optional[asyncio.Task] = None
        self._started_at: str = ""

    async def _exchange_for_token(self) -> str:
        """Exchange passport proof for a short-lived Realtime JWT."""
        import base64

        challenge = os.urandom(32).hex()
        timestamp = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())

        # Sign the challenge with the passport private key
        private_key_bytes = base64.b64decode(self.credentials.private_key_base64)
        signing_key = SigningKey(private_key_bytes)
        signed = signing_key.sign(challenge.encode("utf-8"))
        signature_hex = signed.signature.hex()

        async with httpx.AsyncClient() as client:
            response = await client.post(
                self.config.token_endpoint,
                json={
                    "agent_id": self.credentials.agent_id,
                    "passport_id": self.credentials.passport_id,
                    "challenge": challenge,
                    "signature": signature_hex,
                    "timestamp": timestamp,
                },
            )
            response.raise_for_status()
            data = response.json()
            return data["token"]

    def _sign_nonce(self, nonce: str) -> str:
        """Sign a nonce with the passport private key."""
        import base64

        private_key_bytes = base64.b64decode(self.credentials.private_key_base64)
        signing_key = SigningKey(private_key_bytes)
        signed = signing_key.sign(nonce.encode("utf-8"))
        return signed.signature.hex()

    async def connect(self) -> None:
        """Connect to Supabase Realtime and track Presence."""
        self._started_at = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())

        # Step 1: Exchange passport for Realtime token
        token = await self._exchange_for_token()

        # Step 2: Create Supabase client with the token
        self._supabase = create_client(self.config.supabase_url, token)

        # Step 3: Join channel and track Presence
        channel_name = f"agent:{self.credentials.agent_id}"
        self._channel = self._supabase.channel(channel_name)

        # Listen for verify.request broadcasts
        self._channel.on_broadcast(
            event="verify.request",
            callback=self._handle_verify_request,
        )

        # Subscribe and track presence
        await self._channel.subscribe()
        await self._channel.track({
            "agent_id": self.credentials.agent_id,
            "passport_id": self.credentials.passport_id,
            "passport_fingerprint": self.credentials.passport_fingerprint,
            "sdk": "modei-sdk",
            "sdk_version": VERSION,
            "runtime": "python",
            "env": os.environ.get("MODEI_ENV", "development"),
            "started_at": self._started_at,
        })

        # Step 4: Start token refresh loop
        self._refresh_task = asyncio.create_task(self._refresh_loop())

        # Step 5: Register cleanup handlers
        for sig in (signal.SIGTERM, signal.SIGINT):
            asyncio.get_event_loop().add_signal_handler(sig, self.disconnect)

    async def _refresh_loop(self) -> None:
        """Refresh the Realtime token every 4 minutes."""
        while True:
            await asyncio.sleep(4 * 60)
            try:
                new_token = await self._exchange_for_token()
                if self._supabase:
                    self._supabase.realtime.set_auth(new_token)
            except Exception as e:
                print(f"[modei-realtime] Token refresh failed: {e}", flush=True)

    def _handle_verify_request(self, payload: dict) -> None:
        """Handle verify.request broadcast by signing the nonce."""
        try:
            nonce = payload.get("payload", {}).get("nonce", "")
            if not nonce:
                return

            signature = self._sign_nonce(nonce)

            if self._channel:
                self._channel.send_broadcast(
                    event="verify.response",
                    data={
                        "nonce": nonce,
                        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                        "uptime_seconds": int(time.monotonic()),
                        "passport_fingerprint": self.credentials.passport_fingerprint,
                        "sdk_version": f"modei-sdk@{VERSION}",
                        "signature": signature,
                    },
                )
        except Exception as e:
            print(f"[modei-realtime] Failed to respond to verify challenge: {e}", flush=True)

    def disconnect(self) -> None:
        """Disconnect from Realtime and clean up."""
        if self._refresh_task:
            self._refresh_task.cancel()
            self._refresh_task = None

        if self._channel:
            try:
                self._channel.untrack()
                self._channel.unsubscribe()
            except Exception:
                pass
            self._channel = None

        if self._supabase:
            try:
                self._supabase.remove_all_channels()
            except Exception:
                pass
            self._supabase = None


async def send_heartbeat(
    api_url: str,
    api_key: str,
    agent_id: str,
    passport_id: str,
    sdk_version: str = VERSION,
    runtime: str = "python",
    env: str = "development",
) -> bool:
    """Send a one-time startup heartbeat to the Modei API."""
    async with httpx.AsyncClient() as client:
        try:
            response = await client.post(
                f"{api_url}/api/agents/{agent_id}/heartbeat",
                headers={
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json",
                },
                json={
                    "passport_id": passport_id,
                    "sdk": "modei-sdk",
                    "sdk_version": sdk_version,
                    "runtime": runtime,
                    "env": env,
                    "started_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                },
            )
            return response.is_success
        except Exception:
            return False
